Replaced NewtonSoft with STJ

Eneuman opened the pull request:
This PR is a big one. I realy tried to keep the changes to a minimum.
I have been testing dsc, devagent, restoreagent and endpoint manager for a couple of days and they seem to be working fine. 
I have not tested the routing agent or the envoy pods since I don't know how to test this yet. 

This PR includes the following changes

- [X] Replaces the dependency to NewtonSoft with System.Text.Json
- [X] Moved the files from Common.Json project to the Common project since there should be no need for a seperate project now
- [X] Replaced dependency to Microsoft.AspNetCore.JsonPatch with dependency to SystemTextJsonPatch that has no dependency to NewtonSoft;
- [X] Removed dependency to Microsoft.Rest.* because it has a dependency to NewtonSoft;
- [X] Introduced a typed contract for the communication between endpointmanager and devagent to get around a bug in STJ 6.0

closes #134 
closes #135

Eneuman commented:
@microsoft-github-policy-service agree

elenavillamil commented on src/dsc/dsc.csproj:
```diff
@@ -127,10 +123,7 @@
-        Include="
-            $(MSBuildThisFileDirectory)..\..\NOTICE.txt;
-            $(MSBuildThisFileDirectory)..\..\LICENSE.txt">
+    <None Include="&#xD;&#xA;            $(MSBuildThisFileDirectory)..\..\NOTICE.txt;&#xD;&#xA;            $(MSBuildThisFileDirectory)..\..\LICENSE.txt">
```
was this refactor intended?

Eneuman replied:
No, that was unintentionally.
It has been fixed. Good catch!

elenavillamil reviewed (commented):
Overall looks good! Left a couple minor comments. Before sending to our testing team for validation, I was curious if you could elaborate a bit more on the move from Microsoft.Rest to k8s.Autorest. 

Thank you again for your contribution to this project :)

Eneuman commented:
The goal was to move away from Microsoft.Rest.ClientRuntime since it has a dependency to NTJ.
I also couldn't find a need for the full ClientRuntime. The code only uses the request/reponse wrappers and some extension methods. Theese where already available in the k8s.Autorest namespace so I switch over to theese since they are the same.
The wrappers are only used internally so I could create our own wrappers instead of using the k8s.Autorest namespace if that's prefered?

Eneuman commented:
@microsoft-github-policy-service agree

Eneuman commented on src/common/Exceptions/InvalidUsageException.cs:
```diff
@@ -6,7 +6,7 @@
 using System.Net.Http;
 using Microsoft.BridgeToKubernetes.Common.Logging;
-using Microsoft.Rest;
+using k8s.Autorest;
```
This was done to get access to the wrapper classes used by the Exception constructors.
More details about the decision to switch to the k8s.Autorest namespace is in the comment below :)

Eneuman commented on src/common/EndpointManagement/RequestArguments/EndpointManagerRequestArgument.cs:
```diff
@@ -0,0 +1,14 @@
+    /// <summary>
+    /// 
+    /// </summary>
+    public abstract class EndpointManagerRequestArgument
```
It might be possible but that would require EndpointManagerRequestArgument to be generic like 
`EndpointManagerRequestArgument<T>` so the complexity of the code would increase.

Since some of the request arguments has other properties aswell, I'm not sure if it's worth it.

elenavillamil commented:
> k8s.Autorest

Makes sense. Nope if functionality is equivalent, using k8s.Autorest sounds good. Will work on sending this to our testing team. I will update this PR when I hear back from them.

Eneuman commented:
> > k8s.Autorest
> 
> @Eneuman, makes sense. Nope if functionality is equivalent, using k8s.Autorest sounds good.
> 
> I see the build validation is failing (details below), I can work on fixing it if you cant repo on your end or you dont have the time. It seems mostly just references to deleted csproj files. Just let me know :)
> 
> <img alt="Screenshot 2023-02-02 at 1 50 11 PM" width="1432" src="https://user-images.githubusercontent.com/4663400/216457533-eec2ecca-d833-4649-aeb5-a7f4ae905617.png">

@elenavillamil I fixed the .sln problems but it looks like it still gets a nuget restore error. Can you see what the build server doesn't like ?

elenavillamil commented:
> > > k8s.Autorest
> > 
> > 
> > @Eneuman, makes sense. Nope if functionality is equivalent, using k8s.Autorest sounds good.
> > I see the build validation is failing (details below), I can work on fixing it if you cant repo on your end or you dont have the time. It seems mostly just references to deleted csproj files. Just let me know :)
> 
> @elenavillamil I fixed the .sln problems but it looks like it still gets a nuget restore error. Can you see what the build server doesn't like ?

Hi @Eneuman thank you! Nuget restore error is okay. For security reasons, PRs created from a fork dont have access to run that command. I have run the build pipeline manually and that is green now. Only thing I see left is that there are references in some docker files to "common.json.csproj". Could you remove them? After that we should be good to send this to our testing team :)

Eneuman commented:
> > > > k8s.Autorest
> > > 
> > > 
> > > @Eneuman, makes sense. Nope if functionality is equivalent, using k8s.Autorest sounds good.
> > > I see the build validation is failing (details below), I can work on fixing it if you cant repo on your end or you dont have the time. It seems mostly just references to deleted csproj files. Just let me know :)
> > 
> > 
> > @elenavillamil I fixed the .sln problems but it looks like it still gets a nuget restore error. Can you see what the build server doesn't like ?
> 
> Hi @Eneuman thank you! Nuget restore error is okay. For security reasons, PRs created from a fork dont have access to run that command. I have run the build pipeline manually and that is green now. Only thing I see left is that there are references in some docker files to "common.json.csproj". Could you remove them? After that we should be good to send this to our testing team :)

@elenavillamil The old references in the Docker files has now been removed. Sorry I missed this.

elenavillamil commented:
> > > > > k8s.Autorest
> > > > 
> > > > 
> > > > @Eneuman, makes sense. Nope if functionality is equivalent, using k8s.Autorest sounds good.
> > > > I see the build validation is failing (details below), I can work on fixing it if you cant repo on your end or you dont have the time. It seems mostly just references to deleted csproj files. Just let me know :)
> > > 
> > > 
> > > @elenavillamil I fixed the .sln problems but it looks like it still gets a nuget restore error. Can you see what the build server doesn't like ?
> > 
> > 
> > Hi @Eneuman thank you! Nuget restore error is okay. For security reasons, PRs created from a fork dont have access to run that command. I have run the build pipeline manually and that is green now. Only thing I see left is that there are references in some docker files to "common.json.csproj". Could you remove them? After that we should be good to send this to our testing team :)
> 
> @elenavillamil The old references in the Docker files has now been removed. Sorry I missed this.

@Eneuman thank you! Sorry one last thing I forgot when I reviewed earlier. Could you update deployment/settings/services/imagetag.setting to increase minor version of the images? [example PR](https://github.com/Azure/Bridge-To-Kubernetes/pull/129/files#diff-c8458234d8d537f855b8fc883def8349d194d2835a586ba1305a555f3532c5ce) This will ensure if user is using older image version it will automatically download new image. I think it is worth it since we are doing common changes that can potentially be breaking if images are not updated at the same time as CLI.

elenavillamil commented:
@Eneuman I have sent VSIX to our testing team. If you want to try on your side below are the steps you can follow to build VSIX (it wont let me attach it here due to unsupported extension). If you follow steps below it will use everything from this PR (CLI plus all images)

Clone [vscode extension repo](https://github.com/Azure/vscode-bridge-to-kubernetes)
Update pacakge.json, "expectedCLIVersion" setting to have value: "1.0.20230206.4". 
Run following commands:
   npm i
   npm run compile
   npm run vscode:package
Set bridge environment variable to dev: export BRIDGE_ENVIRONMENT=dev
Open VSCode from this terminal.
Uninstall the extension and install it back from the VSIX generated above.

Eneuman commented:
@elenavillamil Hi!
I'm working on trying to recreate the exception to see whats going on. Could you please post the complete stack trace ?

elenavillamil commented on src/common/Logging/Extensions.cs:
```diff
@@ -118,7 +118,7 @@ public static string Serialize(this object input)
                 return JsonHelpers.SerializeObject(input, serializerSettings);
             }
-            catch (Exception)
+            catch (JsonException)
```
From library log file I see this string "Serialization Error" message [bridge-library-2023-02-07-22-41-06-19390.txt](https://github.com/Azure/Bridge-To-Kubernetes/files/10680038/bridge-library-2023-02-07-22-41-06-19390.txt)

Stack trace below too:
---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Threading.Tasks.Task.GetExceptions(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.GetMemberAndWriteJson(Object obj, WriteStack& state, Utf8JsonWriter writer)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryWrite(Utf8JsonWriter writer, T value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.WriteCoreAsObject(Utf8JsonWriter writer, Object value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.JsonSerializer.WriteUsingSerializer[TValue](Utf8JsonWriter writer, TValue& value, JsonTypeInfo jsonTypeInfo)
   at System.Text.Json.JsonSerializer.Serialize[TValue](Utf8JsonWriter writer, TValue value, JsonSerializerOptions options)
   at Microsoft.BridgeToKubernetes.Common.Logging.Extensions.ExceptionConverter`1.Write(Utf8JsonWriter writer, TExceptionType value, JsonSerializerOptions options)
   at System.Text.Json.Serialization.JsonConverter`1.TryWrite(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.WriteCore(Utf8JsonWriter writer, T& value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.Serialization.JsonConverter`1.WriteCoreAsObject(Utf8JsonWriter writer, Object value, JsonSerializerOptions options, WriteStack& state)
   at System.Text.Json.JsonSerializer.WriteUsingSerializer[TValue](Utf8JsonWriter writer, TValue& value, JsonTypeInfo jsonTypeInfo)
   at System.Text.Json.JsonSerializer.WriteStringUsingSerializer[TValue](TValue& value, JsonTypeInfo jsonTypeInfo)
   at Microsoft.BridgeToKubernetes.Common.Logging.Extensions.Serialize(Object input)
   at System.Linq.Enumerable.SelectArrayIterator`2.ToArray()
   at Microsoft.BridgeToKubernetes.Common.Logging.LoggerBase.FormatMessage(String format, Object[] args)
   at Microsoft.BridgeToKubernetes.Common.Logging.LoggerBase.ExceptionInner(Exception e, EventLevel eventLevel, Boolean handled)
   at Microsoft.BridgeToKubernetes.Common.Logging.LoggerBase.Exception(Exception e, Boolean handled)
   at Microsoft.BridgeToKubernetes.Common.Logging.Log.<>c__DisplayClass19_0.<Exception>b__0(ILogger x)
   at System.Collections.Generic.IEnumerableExtensions.ExecuteForEach[T](IEnumerable`1 items, Action`1 action, CancellationToken cancellationToken)
   at Microsoft.BridgeToKubernetes.Common.Logging.Log._Call(Action`1 action)
   at Microsoft.BridgeToKubernetes.Common.Logging.Log.Exception(Exception e, Boolean handled)
   at Microsoft.BridgeToKubernetes.Common.DevHostAgent.DevHostAgentExecutorClient.<>c__DisplayClass14_1.<<ReversePortForwardStartAsync>b__0>d.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.TrySetCanceled(CancellationToken tokenToRecord, Object cancellationException)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.TrySetCanceled(CancellationToken tokenToRecord, Object cancellationException)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception, Task`1& taskField)
   at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecuteFromThreadPool(Thread threadPoolThread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
--- End of stack trace from previous location ---

elenavillamil replied:
I think because it is throwing TaskCenceledException after this change the catch is no longer catching the exception. I think should go back to generic exception. Although even after that we still need to see why it is trowing to begin with.

hsubramanianaks replied:
I think we can use ExceptionConverters as workaround to solve this issue. Thoughts? we can add a UT to test this if it works or not. Thanks.
```
            try
            {
                serializerSettings.Converters.Add(new ExceptionConverter<Exception>());
                return JsonHelpers.SerializeObject(input, serializerSettings);
            }
            catch (Exception exception)
            {
                //return "Serialization Error";
                return JsonSerializer.Serialize(exception, serializerSettings);
            }

```

Eneuman replied:
Yeah that was my first approach. But I ran into the issue with custom converters being dependent on the same custom converter. The problem occurs when I tried to use `JsonSerializer.Serialize(exception, serializerSettings);` inside the Write function and I'm reusing the serializerSettings parameter from the method. It's a common problem in STJ.
But since Exceptions can have InnerExceptions that can have InnerExceptions and so on, the serializerSettings inside the Write function needs to know how to serializes Exception, but maby there is a way around this?

hsubramanianaks replied:
@Eneuman I think it is because inside the custom coverter you are already serializing the message as JSON and again in the catch block you are doing serialize again might be the issue. May be try using writer.writeString(value.message) and see if it works ?

<img width="380" alt="image" src="https://user-images.githubusercontent.com/105889062/218670410-19be358c-6c1f-4d83-9016-cc6360de0a0e.png">

hsubramanianaks replied:
@Eneuman I tried commenting out the serialize line, it works partially, I will keep looking into and post my findingss soon here.

elenavillamil commented:
> @elenavillamil Hi! I'm working on trying to recreate the exception to see whats going on. Could you please post the complete stack trace ?

Sure, let a comment on the place where i think exception is happening. Added the log file and stack trace. Let me know if you need anything else. I will keep trying to debug as well. Where you able to produce the VSIX with the steps I shared?

Eneuman commented:
@elenavillamil I have been able to reproduce the problem in VS Code but not in VS 2022.

I have found and solved the first problem and that was that the actual error did not show up in the log since when serializing the Exception, a exception was thrown ("Serialize problem").

The meaning of "MaxDepth" in the serializer is totaly different between STJ and NTJ. 
In STJ it means "Throw error if the depth if more then X".
In NTJ it means "Dont read deeper then X when reading JSON".

So I removed MaxDepth since I couldn't find a use for it because we alway want to read all of the JSON in strings.
Now onto finding whats causing the real exception in VS Code.

elenavillamil commented:
> @elenavillamil I have been able to reproduce the problem in VS Code but not in VS 2022.
> 
> I have found and solved the first problem and that was that the actual error did not show up in the log since when serializing the Exception, a exception was thrown ("Serialize problem").
> 
> The meaning of "MaxDepth" in the serializer is totaly different between STJ and NTJ. In STJ it means "Throw error if the depth if more then X". In NTJ it means "Dont read deeper then X when reading JSON".
> 
> So I removed MaxDepth since I couldn't find a use for it because we always want to read all of the JSON in strings. Now onto finding whats causing the real exception in VS Code.
> 
> Can you build a new CLI version please?

@Eneuman good find! So the maxDepth was in common logger. I think it was trying to prevent bridge from logging very big amount of data. Does STJ has other setting that can achieve similar? If we remove that attribute is it possible to do some string manipulation to minimize log size? Okay to leave as is for now, and we can address later if it becomes an issue. Just curious if simple enough to address now.

[Here](https://github.com/Azure/vscode-bridge-to-kubernetes) is the vscode extension code in case you need for debugging. I nee to check if it is possible that json changes broke some contract there.

New CLI version is 20230208.5. I will also build new VSIX from it and attempt to debug this new version as well.

elenavillamil commented:
Forgot to mention, if you want to test local changes without needing a new build, you can build and publish dsc folder locally. then export BRIDGE_BUILD_PATH to point to dsc output directory. For example:
export BRIDGE_BUILD_PATH='/Users/elvilla/projects/Bridge-To-Kubernetes/src/dsc/bin/Debug/netcoreapp3.1/osx-x64/publish/'

Then open VSCode after that has been set. After that you will use bridge binaries from your local build.

Eneuman commented:
@elenavillamil I think I might have found another bug while trying to solve this.
Can you confirm my findings?

This line sets the argument `watch: true`
https://github.com/Azure/Bridge-To-Kubernetes/blob/59c7d82d03668b322d9270f497e31ac7b6521333/src/common/Kubernetes/KubernetesClient.cs#L308

This should result in the Response Content being set to "LineSeperatedHttpContent"
https://github.com/kubernetes-client/csharp/blob/2313b7f6fd0d9659a0454b88546fc3c9e8c8747b/src/KubernetesClient/Kubernetes.cs#L77

and that should result in a Kubernetes Exception "not a watchable request or failed response" when trying to "Watch" it.
https://github.com/kubernetes-client/csharp/blob/d38302d8f160ee706269864cf151c85ed3f8b5b6/src/KubernetesClient/WatcherExt.cs#L40

The strange thing is that it doesnt, but this line throws the exception stead:
https://github.com/Azure/Bridge-To-Kubernetes/blob/59c7d82d03668b322d9270f497e31ac7b6521333/src/common/Kubernetes/KubernetesClient.cs#L192

and thats beacuse `watch: true` is missing.

Setting `watch:true` fixes the warning in the loggfile but I can't figure out why.

Eneuman commented:
@elenavillamil I have found the problem with the exception you received when trying to stop B2K.

The problem is that the serialization of the handled Exception is failing.
Today the code serializes exceptions to get them i json before writing them to log.

Serialization of Exceptions realy should be avoided because of security risks:
https://github.com/dotnet/runtime/issues/31567#issuecomment-558335944

I can see two ways to solve this:
1) Creating a custom serializable object, copy selected property values from the exception to the object and then write that object to the log.
2) Instead of serializing the exception, just use the Exception.ToString() and write the result to log.

I would prefere the second option since that's more future proof.

There is a open issue about making exceptions serializable in the future:
https://github.com/dotnet/runtime/issues/43482

elenavillamil commented:
> ee two ways to solve this:

@Eneuman , thank you for your time debugging this. Unfortunately, while I agree with you that serialization of exception is not ideal, option 2 is not an option at the moment. These logs dont just go to local files, many of them also go to our kusto tables. We have queries and other things that rely on these fields being json. While we can look into refactoring all this to remove the json dependency in the future plus other improvements like less logging, it is not something that would fit into the scope of this work item. What part is causing the exception serialization issue? do you know why it was not happening before?

Eneuman commented:
NewtonSoft serilizes exceptions better then STJ :( 

No worries I'll fix it :)

hsubramanianaks commented on src/common/Extensions/IOperationContextExtensions.cs:
```diff
@@ -3,22 +3,14 @@
 {
     internal static class IOperationContextExtensions
     {
-        public static void SetRequestIds(this IOperationContext context, CloudException ex)
```
Removing this overloaded method was intentional? Is it because it was not used anywhere ?

Eneuman replied:
@hsubramanianaks Yeah it wasn't in use and had "CloudException" as a in parameter so I think this was legacy code from DevSpaces.

hsubramanianaks commented on src/client.sln:
```diff
@@ -88,30 +84,6 @@ Global
 		{713BA354-82CB-45C2-9EDD-ABC1DB6A8442}.Release|x64.Build.0 = Release|Any CPU
 		{713BA354-82CB-45C2-9EDD-ABC1DB6A8442}.Release|x86.ActiveCfg = Release|Any CPU
 		{713BA354-82CB-45C2-9EDD-ABC1DB6A8442}.Release|x86.Build.0 = Release|Any CPU
-		{3A8AEAEF-8F42-4767-B96F-C703F57D3BD0}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
Hi why are these removed any idea. ?

Eneuman commented:
@elenavillamil Hi
I have been trying to reproduce this error in the latest code but everything seems to work fine now.
Can you give it a try and see if you can reproduce it?

If you can reproduce it, I can add extra logging so we can figure out whats going on. 
I fixed a problem with MaxDepth and maby that solved this aswell.

elenavillamil commented:
> @elenavillamil Hi I have been trying to reproduce the "Serailizabe Error" in the latest code but everything seems to work fine now. Can you give it a try and see if you can reproduce it?
> 
> If you can reproduce it, I can add extra logging so we can figure out whats going on. I fixed a problem with MaxDepth and maby that solved this aswell.

@Eneuman , this may have been was my bad. I override docker images in dev environment last week when I started release. I doing a full build with latest version of this PR and I will test in my end and update here.

elenavillamil commented:
> example PR

elenavillamil commented:
> > @elenavillamil Hi I have been trying to reproduce the "Serailizabe Error" in the latest code but everything seems to work fine now. Can you give it a try and see if you can reproduce it?
> > If you can reproduce it, I can add extra logging so we can figure out whats going on. I fixed a problem with MaxDepth and maby that solved this aswell.
> 
> @Eneuman , this may have been was my bad. I override docker images in dev environment last week when I started release. I doing a full build with latest version of this PR and I will test in my end and update here.

Okay I still see this issue. BTW when you add more logging if you can also update image versions on [this file](https://github.com/Azure/Bridge-To-Kubernetes/blob/main/deployment/settings/services/imagetag.setting), it will make it easier when doing kubectl describe to know we are using debhostagent and restoration job images from this PR. It will also prevent it from reusing previous image which maybe incompatible with this change.

Eneuman commented:
> > > @elenavillamil Hi I have been trying to reproduce the "Serailizabe Error" in the latest code but everything seems to work fine now. Can you give it a try and see if you can reproduce it?
> > > If you can reproduce it, I can add extra logging so we can figure out whats going on. I fixed a problem with MaxDepth and maby that solved this aswell.
> > 
> > 
> > @Eneuman , this may have been was my bad. I override docker images in dev environment last week when I started release. I doing a full build with latest version of this PR and I will test in my end and update here.
> 
> @Eneuman Okay I still see this issue (now that dev environment has the images from this change). BTW when you add more logging if you can also update image versions on [this file](https://github.com/Azure/Bridge-To-Kubernetes/blob/main/deployment/settings/services/imagetag.setting), it will make it easier when doing kubectl describe to know we are using debhostagent and restoration job images from this PR. It will also prevent it from reusing previous image which maybe incompatible with this change.

@elenavillamil I have added more temporary logging and updated the image tags. Could you please give it a try and see what exception message you now recive?

You can search the log for "Serialization Error: Debug output".

elenavillamil commented:
> > > > @elenavillamil Hi I have been trying to reproduce the "Serailizabe Error" in the latest code but everything seems to work fine now. Can you give it a try and see if you can reproduce it?
> > > > If you can reproduce it, I can add extra logging so we can figure out whats going on. I fixed a problem with MaxDepth and maby that solved this aswell.
> > > 
> > > 
> > > @Eneuman , this may have been was my bad. I override docker images in dev environment last week when I started release. I doing a full build with latest version of this PR and I will test in my end and update here.
> > 
> > 
> > @Eneuman Okay I still see this issue (now that dev environment has the images from this change). BTW when you add more logging if you can also update image versions on [this file](https://github.com/Azure/Bridge-To-Kubernetes/blob/main/deployment/settings/services/imagetag.setting), it will make it easier when doing kubectl describe to know we are using debhostagent and restoration job images from this PR. It will also prevent it from reusing previous image which maybe incompatible with this change.
> 
> @elenavillamil I have added more temporary logging and updated the image tags. Could you please give it a try and see what exception message you now recive?
> 
> You can search the log for "Serialization Error: Debug output".

@Eneuman library log file no longer has the Serialization Error [bridge-library-2023-02-23-20-16-26-34096.txt](https://github.com/Azure/Bridge-To-Kubernetes/files/10818427/bridge-library-2023-02-23-20-16-26-34096.txt)
. I still see exception on VSCode output window though. I do see stats api comes back to stats-api image, but restore pod linger around for long time with CrashLoopBackOff then it goes away. Here are some more details on that part. I have some time today, so I will spend rest of the afternoon trying to understand what is happening. you can build VSIX with "expectedCLIVersion": "1.0.20230223.3" if you want to run same code i am running (which is new official build from this PR).

2023-02-23T20:26:36.2611625Z | RestorationJob | WARNG | Deployment todo-app/stats-api has already been restored
2023-02-23T20:26:36.2613905Z | RestorationJob | TRACE | Couldn't get agent endpoint
2023-02-23T20:26:36.2617146Z | RestorationJob | TRACE | Event: RestorationJob-AgentPing <json>{"eventName":"RestorationJob-AgentPing","properties":{"RestorePerformed":"false","NumFailedPings":"11","HasConnectedClients":"","Result":"Failed"},"metrics":{"DurationInMs":26}}</json>
2023-02-23T20:26:36.2624526Z | RestorationJob | ERROR | Failed to ping agent 12 times. Exiting...
2023-02-23T20:26:36.3617105Z | RestorationJob | TRACE | AssemblyLoadContext unloading
2023-02-23T20:26:36.3760901Z | RestorationJob | TRACE | Process exiting...
2023-02-23T20:26:36.3765441Z | RestorationJob | TRACE | Process exited
elvilla@Elenas-MacBook-Pro Bridge To Kubernetes % q
zsh: command not found: q
elvilla@Elenas-MacBook-Pro Bridge To Kubernetes % kubectl get pods                                  
NAME                            READY   STATUS             RESTARTS        AGE
database-api-847fcd9684-nbcj4   1/1     Running            0               65m
frontend-598568c849-clknf       1/1     Running            0               65m
stats-api-7f56bcb5c6-p2jcw      1/1     Running            0               11m
stats-api-restore-eaf0d-nzb5k   0/1     CrashLoopBackOff   5 (2m19s ago)   12m
stats-cache-d85ddcfc6-4zk25     1/1     Running            0               65m
stats-queue-74ffdf9c69-wbkgr    1/1     Running            0               65m
stats-worker-6bd69f9cb9-pb7xh   1/1     Running            2 (65m ago)     65m
todos-db-79bdb85468-4lblq       1/1     Running            0

Eneuman commented:
Strange that the serilization error didn't appear. If it do, we should get a good understanding on what's going on from the log. I have been trying to repo the error today but I have been running into other issues so no luck yet but I will continue tomorrow.

elenavillamil commented:
> Strange that the serilization error didn't appear. If it do, we should get a good understanding on what's going on from the log. I have been trying to repo the error today but I have been running into other issues so no luck yet but I will continue tomorrow.
> 
> I do have one idea though. I think the problem might occurs if we try to serialize something that has a Cancelation Token. Cancelation tokens have a handler that contain a property that is not serlizable so I might add a new Converter that skips serilization of Cancelation tokens. They don't contain any useful information for debugging anyways :)

@Eneuman Sounds good! A bit more info on what I am seeing, restorejob is designed to clean up when it sees devhostagent has no connections. Instead it is existing because it fails to talk to devhostagent 12 times (this leads me to believe devhostagent is crashing). Nothing useful on devhostagent logs. Will keep looking.

 ```
if (numFailedPings >= _restorationJobEnvironmentVariables.NumFailedPingsBeforeExit)
                    {
                        _log.Error($"Failed to ping agent {numFailedPings} times. Exiting...");
                        return ExitCode.Fail;
                    }

```

Restore pod real time logs of when bridge ends, it goes from 3 sessions connected to cant reach agent and 2 stats-api pods:
2023-02-23T21:24:50.7681853Z | RestorationJob | TRACE | Pinging agent at 'http://10.244.1.44:50052/api/connectedsessions'
2023-02-23T21:24:50.7699027Z | RestorationJob | TRACE | Agent has 3 connected sessions
2023-02-23T21:24:50.7707426Z | RestorationJob | TRACE | Event: RestorationJob-AgentPing <json>{"eventName":"RestorationJob-AgentPing","properties":{"RestorePerformed":"false","NumFailedPings":"0","HasConnectedClients":"true","Result":"Succeeded"},"metrics":{"DurationInMs":21}}</json>
2023-02-23T21:24:55.7738736Z | RestorationJob | TRACE | Pinging...
2023-02-23T21:24:55.7829051Z | RestorationJob | TRACE | Dependency: Kubernetes <json>{"name":"Kubernetes","target":"GetV1DeploymentAsync","success":true,"duration":null,"properties":{}}</json>
2023-02-23T21:24:55.7922535Z | RestorationJob | TRACE | Dependency: Kubernetes <json>{"name":"Kubernetes","target":"ListPodsInNamespaceAsync","success":true,"duration":null,"properties":{}}</json>
2023-02-23T21:24:55.7925091Z | RestorationJob | TRACE | Dependency: Kubernetes <json>{"name":"Kubernetes","target":"ListPodsForDeploymentAsync","success":true,"duration":null,"properties":{}}</json>
2023-02-23T21:24:55.7933973Z | RestorationJob | WARNG | Found 2 pods for deployment todo-app/stats-api but expected 1
2023-02-23T21:24:55.7937318Z | RestorationJob | TRACE | Couldn't get agent endpoint
2023-02-23T21:24:55.7941284Z | RestorationJob | TRACE | Event: RestorationJob-AgentPing <json>{"eventName":"RestorationJob-AgentPing","properties":{"RestorePerformed":"false","NumFailedPings":"0","HasConnectedClients":"","Result":"Failed"},"metrics":{"DurationInMs":19}}</json>
2023-02-23T21:25:00.7929560Z | RestorationJob | TRACE | Pinging...


Devhostagent logs:

2023-02-23T21:50:45.0529445Z | RemoteAgent | TRACE | Probe request on stream 32
2023-02-23T21:50:45.0531604Z | RemoteAgent | TRACE | Answer with HTTP Okay async.
2023-02-23T21:50:45.0534493Z | RemoteAgent | TRACE | Send 57 via 32
2023-02-23T21:50:45.0557441Z | RemoteAgent | TRACE | ReversePortForwardConnector.HookupStreamData ReadAsync 32 returns 0 bytes.
2023-02-23T21:50:45.0567673Z | RemoteAgent | TRACE | ReversePortForwardConnector.StartReceiveDataAsync finishes stream 32.
2023-02-23T21:50:45.0582036Z | RemoteAgent | TRACE | PortForwardConnector.Disconnect 32
2023-02-23T21:50:45.0593172Z | RemoteAgent | TRACE | AgentHub closed for 80, id 32
2023-02-23T21:50:45.2787459Z | RemoteAgent | TRACE | Shutting down gracefully\nOperation context: <json>{"clientRequestId":"381fdfe3-c230-4eef-93da-8b16adb22ed9","correlationId":"2eb0f5ce-6dc2-492b-8818-7726a4f3670e1677180162125:a14bc30db24d:46a7581205ab","requestId":null,"userSubscriptionId":null,"startTime":"2023-02-23T21:49:40.394476\u002B00:00","userAgent":"RemoteAgent/1.0.20230223.1","requestHttpMethod":null,"requestUri":null,"version":"1.0.20230223.1","requestHeaders":{},"loggingProperties":{"ApplicationName":"RemoteAgent","DeviceOperatingSystem":"Linux 5.4.0-1101-azure #107~18.04.1-Ubuntu SMP Tue Jan 10 21:14:56 UTC 2023","Framework":".NET 6.0.14","ProcessId":1,"TargetEnvironment":"Production"}}</json>
2023-02-23T21:50:45.2867766Z | RemoteAgent | TRACE | AssemblyLoadContext unloading

elenavillamil commented:
> 

It does have probes, but I dont think it is related. It does this pattern:
 
<img width="670" alt="Screenshot 2023-02-23 at 3 51 24 PM" src="https://user-images.githubusercontent.com/4663400/221057997-de4cfd35-d509-408f-b8ef-1a6cfed9bde0.png">

Eneuman commented:
@elenavillamil I think I figured out whats going on.
When you stop debugging, the cancelationToken is set. This causes the DevHostAgentExecutorClient to throw a TaskCanceled exception. See row 102 in function ReversePortForwardStartAsync.

When we try to log the exception we will try to serialize it, but TaskCanceledException contains properties that cant be serialized so it will fail and throw a NotSupportedException.

We only catched the JsonException in the Serialize extansion menthod, so this error leaked all the way up to the gui and henace showed up in VSCode.

I have now changed the logging for theese kinds of exceptions so we still log it in the loggs even if we can serialize it.

elenavillamil commented:
@Eneuman Great! I will do a run of testing on the last version, and if it looks good send off to our testing team. I think I had mentioned that exception leaking earlier. Although I think it may have gotten lost since the conversation was long :) Lets hope it solves it. Will keep you posted.
<img width="955" alt="Screenshot 2023-02-24 at 9 32 17 AM" src="https://user-images.githubusercontent.com/4663400/221248370-5d7d467c-2f7a-4055-829b-62152a51f25c.png">

Eneuman commented:
@elenavillamil Yeah, I missed that comment I think :( It looks like the throw of exception is by design and there is a comment in code to refactor that pattern.

elenavillamil commented:
@Eneuman I can confirm it works now. Sending it off to testing team :)

Eneuman commented:
@elenavillamil yeah, I agree.
I could make a PR just to refactor the "Serilize" extension method since there is now way to add logging to it today. Including the exception message in the return method as I now did feels very hacky.

elenavillamil commented:
Hi @Eneuman , Our testing team reported an issue when isolation mode is enabled. It reproduces on both windows and Mac and I was able to reproduce on my Mac as well. If I have time today I will try to look into it, but it seems it should be easy traceble. It seems serialization issue around WatchEventType. More details below.

Connect shows below error:
```
Preparing to run Bridge To Kubernetes configured as pod todo-app/elvilla-stats-api-7f56bcb5c6-qfgvw ...
Failed to get routing manager deployment status - ran out of time : 'Failed to get routing manager deployment status: 'Invalid value of trigger elvilla-stats-api-7f56bcb5c6-qfgvw'Please include the following Correlation ID when contacting support: '3943796e-75b0-47ba-b1b9-3422a869b93f1677689888147:0feebdf89c3f'. '
Stopping workload and cleaning up...
Restore: Pod 'elvilla-stats-api-7f56bcb5c6-qfgvw' deleted.
```

From library log I only see below, but the pod was in running state when I checked.
```
2023-03-01T17:03:13.4432340Z | Library | WARNG | Routing manager's GetStatus: Routing wasn't setup properly on time: 'Invalid value of trigger elvilla-stats-api-7f56bcb5c6-qfgvw'
2023-03-01T17:03:13.4461940Z | Library | TRACE | Dependency: Kubernetes <json>{"target":"PortForward","success":true,"duration":null,"properties":{"ExitCode":"0"}}</json>
2023-03-01T17:03:15.4469000Z | Library | ERROR | Failed to get routing manager deployment status: 'Invalid value of trigger elvilla-stats-api-7f56bcb5c6-qfgvw'
```

So I checked the pod logs and I see this:
`2023-03-01T17:07:45.5572477Z | RoutingManager | ERROR | Logging handled exception: System.Text.Json.JsonException: {"LineNumber":0,"BytePositionInLine":15,"Path":"$.type","Message":"The JSON value could not be converted to k8s.WatchEventType. Path: $.type | LineNumber: 0 | BytePositionInLine: 15.","Data":{},"InnerException":null,"HelpLink":null,"Source":"System.Text.Json","HResult":-2146233088,"StackTrace":"   at System.Text.Json.ThrowHelper.ThrowJsonException(String message)\n   at System.Text.Json.Serialization.Converters.EnumConverter\u00601.Read(Utf8JsonReader\u0026 reader, Type typeToConvert, JsonSerializerOptions options)\n   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo\u00601.ReadJsonAndSetMember(Object obj, ReadStack\u0026 state, Utf8JsonReader\u0026 reader)\n   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter\u00601.OnTryRead(Utf8JsonReader\u0026 reader, Type typeToConvert, JsonSerializerOptions options, ReadStack\u0026 state, T\u0026 value)\n   at System.Text.Json.Serialization.JsonConverter\u00601.TryRead(Utf8JsonReader\u0026 reader, Type typeToConvert, JsonSerializerOptions options, ReadStack\u0026 state, T\u0026 value)\n   at System.Text.Json.Serialization.JsonConverter\u00601.ReadCore(Utf8JsonReader\u0026 reader, JsonSerializerOptions options, ReadStack\u0026 state)\n   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan\u00601 utf8Json, JsonTypeInfo jsonTypeInfo, Nullable\u00601 actualByteCount)\n   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan\u00601 json, JsonTypeInfo jsonTypeInfo)\n   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)\n   at Microsoft.BridgeToKubernetes.Common.Services.Kubernetes.KubernetesWatcher.ProcessAsync(HttpResponseMessage response, Action\u00602 callback, CancellationToken cancellationToken) in /src/common/Kubernetes/KubernetesWatcher.cs:line 274\n   at Microsoft.BridgeToKubernetes.Common.Services.Kubernetes.KubernetesWatcher.\u003C\u003Ec__DisplayClass17_0.\u003C\u003CWatchAsync\u003Eb__0\u003Ed.MoveNext() in /src/common/Kubernetes/KubernetesWatcher.cs:line 199"}`

Eneuman commented:
@elenavillamil This should now be fixed. 
The problem was that STJ does not support EnumMember out of the box so I had to build a new JsonConverter. I also added a test for this issue.

elenavillamil commented on src/common/Models/Settings/PortPair.cs:
```diff
@@ -47,19 +47,19 @@ public PortPair(int remotePort, string protocol = KubernetesConstants.Protocols.
+        [JsonPropertyName("protocol")]
         public string Protocol { get; set; }
 
         /// <summary>
```
This is new property added by recent PR, we need to update it from JsonProperty to JsonPropertyName as well. If I have time today i can try to push the change myself.

elenavillamil replied:
updated the pull request (hope its okay, i just didnt want to bother you for a small change needed after merging with main)

elenavillamil reviewed (approved):
LGMT. Testing team reported passings tests as well. Thank you for this change @Eneuman !
